src/serial.rs

```diff
@@ -413,52 +517,168 @@ macro_rules! hal {
                 }
             }
 
-            impl<PINS> crate::hal::serial::Read<u8> for Serial<$USARTX, PINS> {
+            impl crate::hal::serial::Write<u16> for Tx<$USARTX, u16> {
+                type Error = Infallible;
+
+                fn flush(&mut self) -> nb::Result<(), Self::Error> {
+                    unsafe { &*$USARTX::ptr() }.flush()
+                }
+                fn write(&mut self, word: u16) -> nb::Result<(), Self::Error> {
+                    unsafe { &*$USARTX::ptr() }.write16(word)
+                }
+            }
+
+            impl<PINS> crate::hal::serial::Read<u8> for Serial<$USARTX, PINS, u8> {
                 type Error = Error;
 
                 fn read(&mut self) -> nb::Result<u8, Error> {
-                    self.usart.deref().read()
+                    self.usart.deref().read8()
+                }
+            }
+
+            impl<PINS> crate::hal::serial::Read<u16> for Serial<$USARTX, PINS, u16> {
+                type Error = Error;
+
+                fn read(&mut self) -> nb::Result<u16, Error> {
+                    self.usart.deref().read16()
                 }
             }
 
-            impl<PINS> crate::hal::serial::Write<u8> for Serial<$USARTX, PINS> {
+            impl<PINS> crate::hal::serial::Write<u8> for Serial<$USARTX, PINS, u8> {
                 type Error = Infallible;
 
                 fn flush(&mut self) -> nb::Result<(), Self::Error> {
                     self.usart.deref().flush()
                 }
 
                 fn write(&mut self, byte: u8) -> nb::Result<(), Self::Error> {
-                    self.usart.deref().write(byte)
+                    self.usart.deref().write8(byte)
                 }
             }
 
+            impl<PINS> crate::hal::serial::Write<u16> for Serial<$USARTX, PINS, u16> {
+                type Error = Infallible;
+
+                fn flush(&mut self) -> nb::Result<(), Self::Error> {
+                    self.usart.deref().flush()
+                }
+
+                fn write(&mut self, word: u16) -> nb::Result<(), Self::Error> {
+                    self.usart.deref().write16(word)
+                }
+            }
         )+
     }
 }
 
-impl<USART> core::fmt::Write for Tx<USART>
+impl<USARTX, PINS, WORD> Serial<USARTX, PINS, WORD> {
+    /// Returns ownership of the borrowed register handles
+    pub fn release(self) -> (USARTX, PINS) {
+        (self.usart, self.pins)
+    }
+
+    /// Separates the serial struct into separate channel objects for sending (Tx) and
+    /// receiving (Rx)
+    pub fn split(self) -> (Tx<USARTX, WORD>, Rx<USARTX, WORD>) {
+        (
+            Tx {
+                _usart: PhantomData,
+                _word: PhantomData,
+            },
+            Rx {
+                _usart: PhantomData,
+                _word: PhantomData,
+            },
+        )
+    }
+}
+
+impl<USARTX, PINS> Serial<USARTX, PINS, u8> {
+    pub fn as_u16(self) -> Serial<USARTX, PINS, u16> {
```

Comment:
This name doesn't say anything.

Reply:
I couldn't think of anything better. Can you suggest your own version?

Reply:
`into_9bit_word_serial` for example

Reply:
I think this name can be confusing. This function does not affect the world length of the serial port in any way. This function only allows you to change the bit depth of access to the data register (USART_DR).

Reply:
Then something like `with_u16_data`

Reply:
When would you want this function if it doesn't change the word length? Reading 9 bit data into a u8 seems pointless, as does reading 8 bit data into a u16

Reply:
The `serial` is always created as u8 data regardless of the configuration (wordlength). Then, if necessary, it can be switched to u16 data.